src/utils/leb128.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+export interface Leb128DecodeResult {
+  value: bigint;
+  rest: Buffer;
+}
+
+export function encodeSigned(value: bigint | number, maxBytes: number | null = null): Buffer {
+  let val = BigInt(value);
+  const result: bigint[] = [];
+  while (true) {
+    const byte = val & 0b0111_1111n;
+    val >>= 7n;
+    if (
+      (val === 0n && (byte & 0b0100_0000n) === 0n) ||
+      (val === -1n && (byte & 0b0100_0000n) !== 0n)
+    ) {
+      result.push(byte);
+      if (maxBytes !== null && result.length > maxBytes) {
+        throw new Error(`Cannot encode more than ${maxBytes} bytes`);
+      }
+      // Need to convert the bigint values of `result` into Number to use with Buffer.
+      // This is a safe operation since the value of the elements can only go up to 0xFF
+      return Buffer.from(result.map(b => Number(b)));
+    }
+
+    // Add 7 bits + first bit indicating the existence of the next block.
+    result.push(byte | 0b1000_0000n);
+    if (maxBytes !== null && result.length > maxBytes) {
+      throw new Error(`Cannot encode more than ${maxBytes} bytes`);
+    }
+  }
+}
+
+export function decodeSigned(buf: Buffer, maxBytes: number | null = null): Leb128DecodeResult {
+  const byte_list = Array.from(buf.values()).map(v => BigInt(v));
+  let result = 0n;
+  let shift = 0n;
+  while (true) {
+    const byte = byte_list.shift();
+    if (byte === undefined) {
+      throw new Error('Buffer is not valid leb128, cannot read from empty buffer');
+    }
+    result |= (byte & 0b0111_1111n) << shift;
+    shift += 7n;
+    // assert shift % 7n === 0
+    if (shift % 7n !== 0n)
+      throw new Error(`AssertionError: shift is ${shift} and is not divisible by 7`);
+
+    if (maxBytes !== null && shift / 7n > maxBytes) {
```

Comment:
In the python implementation it's using `shift // 7`. Are there any reasons not to do the integer division?

Reply:
Ahh ok! I think you should just leave a comment in the code that this is already an integer division because it's a bigint.